adding mfa code verify and adding trusted token

SlemanAlrstum231 opened the pull request.

AlrstumSleman committed:
adding mfa code verify and adding trusted token

AlrstumSleman committed:
adding project name checker befor adding the project

sarmadka commented on Drivers/AlususNet.alusus:
```diff
@@ -71,314 +85,416 @@
+            promptForUserName[];
+            if Fs.exists(this.trustedTokenPath) {
+                this.trustKey = Fs.readFile(this.trustedTokenPath);
+            }
```
لنخزن اسم المستخدم مع الtrustToken ونقارنه هنا.

sarmadka commented on Drivers/AlususNet.alusus:
```diff
@@ -71,314 +85,416 @@
+                        request.post(String.format(
+                            "{\"username\": \"%s\"}",
+                            Json.escape(String(userName~ptr).trim()).buf
+                        ));
```
لنضف كلمة السر إلى هذا الطلب أيضًا

sarmadka replied:
لنخزن عنوان IP ضمن اسم توكن الMFA، ثم نقارنه لاحقًا عند مقارنة الكود.

sarmadka commented on Drivers/AlususNet.alusus:
```diff
@@ -71,314 +85,416 @@
+                            promptForMfaCode[];
+                            def verifyMfaCodeRequest: Net.Request(host + CREATE_AUTH_TOKEN_MFA);
+                            verifyMfaCodeRequest.post(String.format(
+                                "{\"code\": \"%s\"}",
```
لنضف اسم المستخدم إلى هذا الطلب.

sarmadka commented on Drivers/AlususNet.alusus:
```diff
@@ -71,314 +85,416 @@
+                                    Console.print(I18n.string("Code is invalid!\n"));
+                                    if this.invalidCodeCounter > 0 {
+                                        this.invalidCodeCounter--;
+                                        promptForMfaCode[];
```
في هذه الحالة نحتاج للرجوع إلى نقطة مقارنة الكود بدل نقطة إرسال كود جديد. ونعطي للمستخدم خيار طلب إرسال كود جديد إن أراد.

sarmadka commented on Drivers/AlususNet.alusus:
```diff
@@ -71,314 +85,416 @@
+                    }
                 }
+                def data : String = String.format("{\"authKey\": \"%s\"}",this.authKey.buf);
+                Fs.createFile(this.settingsFilePath, data, data.getLength());
```
لا داعي لتخزين الauthToken في ملف

sarmadka commented on Drivers/AlususNet.alusus:
```diff
@@ -71,314 +85,416 @@
+            handler this.fetchUploadToken() {
+                if Fs.exists(this.settingsFilePath) {
+                    def data : String = Fs.readFile(this.settingsFilePath);
+                    this.authKey = Json(data)("authKey");
```
في هذه الحالة نقرأ الuploadKey من الملف ونرجع من الدالة مباشرة.

AlrstumSleman committed:
fixing mfa code with the new updates on Madeef

sarmadka committed:
Merge remote-tracking branch 'origin/main' into adding_mfa_verfiy

sarmadka committed:
Fixed issue caused by merge conflict.

sarmadka committed:
Pick asset routes from asset endpoints.

* Extract asset paths from asset route definitions in the
  source code instead of having to manually provide them
  in the config.
* Added `initializer` optional propert to `Config`.
* Enable optionally providing a different name for
  publishing in AlususNetPublishDriver.
* Added missing translations.
